### PR TITLE
feat(RichTextEditor): format toolbar [WPB-12089]

### PR DESCRIPTION
**Pull request by olafsulich**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-12089" title="WPB-12089" target="_blank">StoryWPB-12089</a>  [Web] Text formatting via UI in text input field
  </td></table>
  <br />
 

## Description

A new rich text editor that allows users to edit their message visually before sending it, contains options: heading (h1), bold, italic, strikethrough, lists (ordered & unordered), and code.

[Figma Designs](https://www.figma.com/design/jnygr7PEOk1yHgjBF18udc/Desktop?node-id=4530-86442&node-type=section&t=KutImvdm0QGOcHrB-0)

**What's changed?**
- added the format toolbar which activates after the "show formatting" button, it's behind a flag
- added the emoji picker modal, same as for message reactions, also behind a flag
- added the strikethrough option for the markdown renderer, change how inline code is displayed (with background)
- small refactoring, moved files around, created some components
- changed avatar to be visible all the time (behind flag), not only when typing
- changed the order of action buttons (pick image/file, ping, etc.) as it's on design

**What's missing?**

Tests. It turned out, that testing the Lexical-related code isn't simple. It's because of the poor support for the `contenteditable` in JSDOM (issue [no.1](https://github.com/jsdom/jsdom/issues/1670), and [no.2](https://github.com/facebook/lexical/issues/4595)), and lack of other utilities from the Lexical itself, [more info](https://github.com/facebook/lexical/discussions/2659). We will add end-to-end tests with QA guidance to ensure the feature works properly.

## Screenshots/Screencast (for UI changes)

**Text formatting:**

https://github.com/user-attachments/assets/db8c1dcc-b931-412a-8634-9e7771eac4f0

**Message editing:**

https://github.com/user-attachments/assets/dc5edc2b-df54-4b2d-bedd-56fec17fc122

**Message editing (rich, long):**

https://github.com/user-attachments/assets/eace32de-98c8-443c-ae18-1da8808b3353

**Inserting emoji:**

https://github.com/user-attachments/assets/9792b3d3-86b4-4a94-b94f-3baef6a5342d

**Reply to message preview**:

https://github.com/user-attachments/assets/1b299d8c-1205-412b-87c2-cd5b57140b45

--- 

**Default view, large viewport**:

image

**Default view, small viewport**:

image

**Rich text view, large viewport**:

image

**Rich text view, small viewport**:

image

## Checklist

- [x] mentions the JIRA issue in the PR name (Ex. [WPB-423])
- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [x] If it is a core feature, unit tests have been added;

[WPB-423]: https://wearezeta.atlassian.net/browse/WPB-423?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/script/components/InputBar/InputBar.tsx [hunk 1]
```diff
@@ -534,107 +548,113 @@ export const InputBar = ({
...
     showGiphyButton: showGiphyButton,
+    isFormatActive: formatToolbar.open,
+    onFormatClick: formatToolbar.handleClick,
+    isEmojiActive: emojiPicker.open,
+    onEmojiClick: emojiPicker.handleToggle,
   };
 
   const enableSending = textValue.length > 0;
 
+  const showAvatar = messageFormatButtonsEnabled ? true : !!textValue.length;
```
Comment (przemvs) [inline src/script/components/InputBar/InputBar.tsx]:
```suggestion
  const showAvatar = messageFormatButtonsEnabled || !!textValue.length;
```

File: src/script/components/InputBar/components/ImageUploadButton/ImageUploadButton.test.tsx [hunk 2]
```diff
@@ -30,7 +30,11 @@ describe('ImageUploadButton', () => {
     const onSelectImages = jest.fn();
 
     const {container} = render(
-      <ImageUploadButton onSelectImages={onSelectImages} acceptedImageTypes={ALLOWED_IMAGE_TYPES} />,
+      <ImageUploadButton
+        onSelectImages={onSelectImages}
+        acceptedImageTypes={ALLOWED_IMAGE_TYPES}
+        hasRoundedCorners={true}
```
Comment (przemvs) [inline src/script/components/InputBar/components/ImageUploadButton/ImageUploadButton.test.tsx]:
```suggestion
        hasRoundedCorners
```

File: src/script/components/InputBar/components/ImageUploadButton/ImageUploadButton.test.tsx [hunk 3]
```diff
@@ -45,7 +49,11 @@ describe('ImageUploadButton', () => {
...
```
Comment (przemvs) [inline src/script/components/InputBar/components/ImageUploadButton/ImageUploadButton.test.tsx]:
```suggestion
        hasRoundedCorners
```

File: src/script/components/RichTextEditor/components/FormatToolbar/useHeadingState/useHeadingState.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import {isNodeHeading} from '../common/isNodeHeading/isNodeHeading';
+
+const INSERT_HEADING_COMMAND = createCommand();
+
+export const useHeadingState = () => {
+  const [editor] = useLexicalComposerContext();
+
+  const registerHeadingCommand = useCallback(() => {
+    editor.registerCommand(INSERT_HEADING_COMMAND, headingCommand, 0);
+  }, [editor]);
```
Comment (przemvs) [inline src/script/components/RichTextEditor/components/FormatToolbar/useHeadingState/useHeadingState.ts]: I think You can move it into useEffect and use return editor.registerCommand ;)

File: src/script/components/InputBar/components/ImageUploadButton/ImageUploadButton.test.tsx (hunk 2 above)
Comment (olafsulich) [reply src/script/components/InputBar/components/ImageUploadButton/ImageUploadButton.test.tsx]: habit from previous project 😅 going to fix that

File: src/script/components/RichTextEditor/components/FormatToolbar/useToolbarState/useToolbarState.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  }, [editor]);
+
+  useEffect(() => {
+    if (!editor) {
+      return undefined;
+    }
+
+    const unregister = editor.registerUpdateListener(updateToolbar);
+
+    return unregister;
```
Comment (przemvs) [inline src/script/components/RichTextEditor/components/FormatToolbar/useToolbarState/useToolbarState.ts]: I think we don't have to passing it to variable ( redutant code ). So let's go return it directly ;)

File: src/script/components/InputBar/InputBar.tsx [hunk 6]
```diff
@@ -534,107 +548,113 @@ export const InputBar = ({
...
+                  showFormatToolbar={formatToolbar.open}
+                  saveDraftState={saveDraft}
+                  loadDraftState={loadDraft}
+                  onShiftTab={onShiftTab}
+                  onSend={handleSendMessage}
+                  onBlur={() => isTypingRef.current && conversationRepository.sendTypingStop(conversation)}
+                >
+                  <ul
+                    className={cx('controls-right buttons-group input-bar-buttons', {
+                      'controls-right-shrinked': textValue.length !== 0,
```
Comment (thisisamir98) [inline src/script/components/InputBar/InputBar.tsx:632]:
```suggestion
                      'controls-right-shrinked': !!textValue.length,
```

File: src/script/components/InputBar/InputBar.tsx [hunk 7]
```diff
@@ -534,107 +548,113 @@ export const InputBar = ({
...
+                  onSetup={lexical => {
+                    editorRef.current = lexical;
+                  }}
+                  editedMessage={editedMessage}
+                  onEscape={() => {
+                    if (editedMessage) {
+                      cancelMessageEditing(true);
+                    } else if (replyMessageEntity) {
+                      cancelMessageReply();
+                    }
```
Comment (thisisamir98) [inline src/script/components/InputBar/InputBar.tsx:611]: maybe extract to a function

File: src/script/components/InputBar/components/InputBarControls/FormatTextButton/FormatTextButton.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+interface FormatTextButtonProps {
+  isActive: boolean;
+  onClick: (event: MouseEvent<HTMLButtonElement>) => void;
+}
+
+export const FormatTextButton = ({isActive, onClick}: FormatTextButtonProps) => {
+  return (
+    <button
+      className={cx(`controls-right-button buttons-group-button-left`, {
```
Comment (thisisamir98) [inline src/script/components/InputBar/components/InputBarControls/FormatTextButton/FormatTextButton.tsx:35]:
```suggestion
      className={cx('controls-right-button buttons-group-button-left', {
```

File: src/script/components/InputBar/hooks/useEmojiPicker/useEmojiPicker.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  const handleToggle = (event: MouseEvent<HTMLButtonElement>) => {
+    const rect = event.currentTarget.getBoundingClientRect();
+    // eslint-disable-next-line id-length
+    emojiPickerPosition.current = {x: rect.x, y: rect.y};
+    setOpen(prev => !prev);
+  };
+
+  useClickOutside(wrapperRef, () => {
+    if (open) {
+      setOpen(false);
```
Comment (przemvs) [inline src/script/components/InputBar/hooks/useEmojiPicker/useEmojiPicker.ts]: You can reuse here handleClose :)

File: src/script/components/RichTextEditor/components/FormatToolbar/FormatToolbar.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+export const FormatToolbar = () => {
+  const [editor] = useLexicalComposerContext();
+
+  const {activeFormats} = useToolbarState();
+
+  const {toggleHeading} = useHeadingState();
+
+  const {toggleList} = useListState();
+
+  const formatText = (format: 'bold' | 'italic' | 'strikethrough' | 'code') => {
```
Comment (przemvs) [inline src/script/components/RichTextEditor/components/FormatToolbar/FormatToolbar.tsx]: Maybe You can use pick from useToolbarState types ? 🤔  From this one: FormatTypes

**Review by thisisamir98 — APPROVED** [review]

**Review by przemvs — APPROVED** [review]

**Pull request merged**